shell/platform/windows/accessibility_root_node.cc

```diff
@@ -0,0 +1,271 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/windows/accessibility_root_node.h"
+
+#include "flutter/fml/logging.h"
+
+namespace flutter {
+
+static constexpr LONG kWindowChildId = 1;
+static constexpr LONG kInvalidChildId = 3;
+
+AccessibilityRootNode::AccessibilityRootNode() : alert_accessible_(nullptr) {}
+
+AccessibilityRootNode::~AccessibilityRootNode() {
+  if (alert_accessible_) {
+    alert_accessible_->Release();
+    alert_accessible_ = nullptr;
+  }
+}
+
+IAccessible* AccessibilityRootNode::GetTargetAndChildID(VARIANT* var_id) {
+  LONG& child_id = var_id->lVal;
+  if (V_VT(var_id) != VT_I4) {
+    child_id = kInvalidChildId;
+    return FALSE;
+  }
+  child_id = V_I4(var_id);
+  if (!window_accessible_) {
+    return nullptr;
+  }
+  if (child_id == CHILDID_SELF || child_id == kWindowChildId) {
+    child_id = CHILDID_SELF;
+    return window_accessible_;
+  }
+  if (child_id == kAlertChildId && alert_accessible_) {
+    child_id = CHILDID_SELF;
+    return alert_accessible_;
+  }
+  if (child_id < 0) {
+    return window_accessible_;
+  }
+  return nullptr;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::accHitTest(LONG screen_physical_pixel_x,
+                                                 LONG screen_physical_pixel_y,
+                                                 VARIANT* child) {
+  if (window_accessible_) {
+    return window_accessible_->accHitTest(screen_physical_pixel_x,
+                                          screen_physical_pixel_y, child);
+  }
+  child->vt = VT_EMPTY;
+  return S_FALSE;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::accDoDefaultAction(VARIANT var_id) {
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&var_id))) {
+    return target->accDoDefaultAction(var_id);
+  }
+  return E_FAIL;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::accLocation(LONG* physical_pixel_left,
+                                                  LONG* physical_pixel_top,
+                                                  LONG* width,
+                                                  LONG* height,
+                                                  VARIANT var_id) {
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&var_id))) {
+    return target->accLocation(physical_pixel_left, physical_pixel_top, width,
+                               height, var_id);
+  }
+  return S_FALSE;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::accNavigate(LONG nav_dir,
+                                                  VARIANT start,
+                                                  VARIANT* end) {
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&start))) {
+    return target->accNavigate(nav_dir, start, end);
+  }
+  return S_FALSE;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::get_accChild(VARIANT var_child,
+                                                   IDispatch** disp_child) {
```

Comment:
Could this use `GetTargetAndChildID`?

Reply:
`GetTargetAndChildID` returns `window_accessible_` when `var_child.lVal == CHILDID_SELF`, but this method should return `this` in that case. It would be possible to use, but it would need additional logic for cases like this which may counteract the benefit from calling it here.

Reply:
Gotcha, makes sense 👍